### PR TITLE
🏃 add tests for gcp_machine_controller 

### PR DESCRIPTION
**Pull request by nader-ziada**
**What this PR does / why we need it**:
- improve test coverage
- create a mock gcp compute service to be able to mock out the calls to gcp in the test
- change the controller to allow for passing of the mock service

### CODE REVIEW

File: controllers/gcpmachine_controller_test.go [hunk 1]
```diff
@@ -17,12 +17,27 @@ limitations under the License.
 package controllers
 
 import (
+	"bytes"
+	"context"
+	"flag"
+
+	capierrors "sigs.k8s.io/cluster-api/errors"
```
Comment (detiber) [inline controllers/gcpmachine_controller_test.go]: nit, this should be grouped with the imports below

File: controllers/gcpmachine_controller_test.go [hunk 2]
```diff
@@ -48,4 +64,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+			machineScope.SetErrorReason(capierrors.UpdateMachineError)
+			machineScope.SetErrorMessage(errors.New("GCE error"))
+			Expect(k8sClient.Update(ctx, machineScope.GCPMachine)).Should(Succeed())
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Error state detected, skipping reconciliation"))
```
Comment (detiber) [inline controllers/gcpmachine_controller_test.go]: Checking for a specific log message seems a bit brittle. Do we need a better way to bubble this information up?

File: controllers/gcpmachine_controller_test.go [hunk 3]
```diff
@@ -48,4 +64,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+		It("should return if infrastructure is not ready", func() {
+			clusterStatus := clusterv1.ClusterStatus{InfrastructureReady: false}
+			machineScope.Cluster.Status = clusterStatus
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Cluster infrastructure is not ready yet"))
```
Comment (detiber) [inline controllers/gcpmachine_controller_test.go]: Checking for a specific log message seems a bit brittle. Do we need a better way to bubble this information up?

File: controllers/gcpmachine_controller_test.go [hunk 4]
```diff
@@ -48,4 +64,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+					Data: nil,
+				}}
+			machineScope.Machine.Spec = spec
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Bootstrap data is not yet available"))
```
Comment (detiber) [inline controllers/gcpmachine_controller_test.go]: Checking for a specific log message seems a bit brittle. Do we need a better way to bubble this information up?

File: controllers/gcpmachine_controller_test.go (hunk 1 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: will change

File: controllers/gcpmachine_controller_test.go (hunk 2 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: The controller currently sets the machineScope.info with this message and intentionally sets the error to nil, would you prefer returning a specific error?

Comment (detiber) [reply controllers/gcpmachine_controller_test.go]: I believe there was an issue with returning an error and having a way to tell controller-runtime not to requeue the request. According to https://github.com/kubernetes-sigs/controller-runtime/issues/617 it looks like this is still an outstanding issue.

File: controllers/gcpmachine_controller_test.go [hunk 5]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+		})
+
+		AfterEach(func() {
+			Expect(k8sClient.Delete(ctx, machineScope.GCPMachine)).Should(Succeed())
+		})
+
+		It("should return early if gcp machine is in error state", func() {
+			machineScope.SetErrorReason(capierrors.UpdateMachineError)
+			machineScope.SetErrorMessage(errors.New("GCE error"))
+			Expect(k8sClient.Update(ctx, machineScope.GCPMachine)).Should(Succeed())
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Do we need this Update here?

File: controllers/gcpmachine_controller_test.go [hunk 6]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+						Name:       gcpMachine.Name,
+					},
+					Bootstrap: clusterv1.Bootstrap{
+						Data: &bootstrapData,
+					},
+					Version: &version,
+				},
+			}
+
+			Expect(k8sClient.Create(ctx, gcpMachine)).Should(Succeed())
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Is there a need to use `k8sClient`? Can we just set up a `fakeClient = fake.NewFakeClient()` and use that everywhere?

File: controllers/gcpmachine_controller_test.go [hunk 7]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+					Namespace: "default",
+				},
+				Spec: infrav1.GCPClusterSpec{
+					Project: "gcp-project",
+				},
+			}
+
+			// Create the cluster scope
+			clusterScope, err = scope.NewClusterScope(scope.ClusterScopeParams{
+				Client:     fake.NewFakeClient(),
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Use the above `fakeClient`?

File: controllers/gcpmachine_controller_test.go [hunk 8]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+				Logger:     log.Log,
+				Cluster:    cluster,
+				GCPCluster: gcpCluster,
+			})
+			Expect(err).To(BeNil())
+
+			// Create the machine scope
+			machineScope, err = scope.NewMachineScope(scope.MachineScopeParams{
+				Logger:     log.Log,
+				Client:     fake.NewFakeClient(),
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Use the above `fakeClient`?

File: controllers/gcpmachine_controller_test.go [hunk 9]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+				Machine:    machine,
+				GCPCluster: gcpCluster,
+				GCPMachine: gcpMachine,
+			})
+			Expect(err).To(BeNil())
+
+		})
+
+		AfterEach(func() {
+			Expect(k8sClient.Delete(ctx, machineScope.GCPMachine)).Should(Succeed())
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Use the above `fakeClient`?

File: controllers/gcpmachine_controller_test.go [hunk 10]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+			machineScope, err = scope.NewMachineScope(scope.MachineScopeParams{
+				Logger:     log.Log,
+				Client:     fake.NewFakeClient(),
+				Cluster:    cluster,
+				Machine:    machine,
+				GCPCluster: gcpCluster,
+				GCPMachine: gcpMachine,
+			})
+			Expect(err).To(BeNil())
+
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: nit: extra unnecessary whitespace

File: controllers/gcpmachine_controller_test.go [hunk 11]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+			clusterStatus := clusterv1.ClusterStatus{InfrastructureReady: false}
+			machineScope.Cluster.Status = clusterStatus
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Cluster infrastructure is not ready yet"))
+
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: nit: extra unnecessary whitespace

File: controllers/gcpmachine_controller_test.go [hunk 12]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+				}}
+			machineScope.Machine.Spec = spec
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Bootstrap data is not yet available"))
+
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: nit: extra unnecessary whitespace

File: controllers/gcpmachine_controller_test.go [hunk 13]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+
+			By("Checking providerId")
+			Expect(machineScope.GetProviderID()).To(Equal("gce://gcp-project/us-west4-a/gce-instance"))
+			By("Checking instanceStatus")
+			Expect(*machineScope.GetInstanceStatus()).To(Equal(infrav1.InstanceStatus("RUNNING")))
+			By("Checking a finalizer is added")
+			Expect(machineScope.GCPMachine.Finalizers).To(ContainElement(infrav1.MachineFinalizer))
+
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: nit: extra unnecessary whitespace

File: controllers/gcpmachine_controller_test.go [hunk 14]
```diff
@@ -48,4 +63,171 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+				Spec: clusterv1.MachineSpec{
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
+						Kind:       "GCPMachine",
+						Name:       gcpMachine.Name,
+					},
+					Bootstrap: clusterv1.Bootstrap{
+						Data: &bootstrapData,
+					},
+					Version: &version,
```
[the remark this answers is not in the document]
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: This is a required field that will give a validation error if not present

File: controllers/gcpmachine_controller_test.go (hunk 6 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: k8sclient is setup in the `suite_test` and is the recommended way in kubebuilder docs. I was already there as part of the existing test

File: controllers/gcpmachine_controller_test.go (hunk 7 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: yes, good catch, no need for another fake client, will use k8sClient

File: controllers/gcpmachine_controller_test.go (hunk 8 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: will switch to k8sClient

File: controllers/gcpmachine_controller_test.go (hunk 10 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: ok, will remove the extra lines

File: controllers/gcpmachine_controller_test.go (hunk 9 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: k8sClient is the client setup in suite_test, removed the other fake client

File: controllers/gcpmachine_controller_test.go (hunk 5 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: I guess it can be removed

File: controllers/gcpmachine_controller_test.go (hunk 2 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: there is a series of checks that happen before this call is made, so I don't think just checking for `InstanceIfExists` not being called would be enough replacement that the right check failed

File: controllers/gcpmachine_controller_test.go (hunk 3 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: there is a series of checks that happen before this call is made, so I don't think just checking for InstanceIfExists not being called would be enough replacement that the right check failed

File: controllers/gcpmachine_controller_test.go (hunk 11 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: ok

File: controllers/gcpmachine_controller_test.go (hunk 4 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: there is a series of checks that happen before this call is made, so I don't think just checking for InstanceIfExists not being called would be enough replacement that the right check failed

File: controllers/gcpmachine_controller_test.go (hunk 12 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: ok

File: controllers/gcpmachine_controller_test.go (hunk 13 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: ok

**Comment by nader-ziada** [comment]
@detiber @wfernandes since you looked at this code, can one of you give this PR an `ok-to-test` if you think it makes sense. Thanks

**Comment by nader-ziada** [comment]
I've had a chat with @wfernandes and explained why checking that a func has not been called yet in the test might not guarantee that the required validation happened since all these validation happen in sequence in the same `reconcile` method. 

is there anything else I should be changing here? 

Thanks

File: controllers/gcpmachine_controller_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (wfernandes) [reply controllers/gcpmachine_controller_test.go]: Hmm...what validation is this? When I take it out and run `make test` everything passes. And currently the property is optional so I'm confused as to why it is required in this context. (https://github.com/kubernetes-sigs/cluster-api/blob/900a2ff1f6688773ff48dc56c459484f22c21c6b/api/v1alpha3/machine_types.go#L60-L61).

[the remark this answers is not in the document]
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: can't find this check anymore, will remove it

File: controllers/gcpmachine_controller_test.go [hunk 15]
```diff
@@ -48,4 +62,163 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+				GCPMachine: gcpMachine,
+			})
+			Expect(err).To(BeNil())
+		})
+
+		AfterEach(func() {
+			Expect(k8sClient.Delete(ctx, machineScope.GCPMachine)).Should(Succeed())
+		})
+
+		It("should return early if gcp machine is in error state", func() {
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Can we change the test title to better reflect what is being tested? Currently it doesn't seem like we are testing early return. Is there a way to assert that reconciliation hasn't happened? Maybe verifying if ProviderID isn't set. since that is a mandatory step in reconciliation. WDYT?

File: controllers/gcpmachine_controller_test.go [hunk 16]
```diff
@@ -48,4 +62,163 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Error state detected, skipping reconciliation"))
+		})
+
+		It("should return if infrastructure is not ready", func() {
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Same comment as above.

File: controllers/gcpmachine_controller_test.go [hunk 17]
```diff
@@ -48,4 +62,163 @@ var _ = Describe("GCPMachineReconciler", func() {
...
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Cluster infrastructure is not ready yet"))
+		})
+
+		It("should return if bootstrap data is not available", func() {
```
Comment (wfernandes) [inline controllers/gcpmachine_controller_test.go]: Same comment as above.

File: controllers/gcpmachine_controller_test.go (hunk 15 above)
Comment (nader-ziada) [reply controllers/gcpmachine_controller_test.go]: I'm not sure I understand your point? returning early is just another way of saying reconciliation didn't happen and since there is an error message it also means reconciliation didn't happen

**Comment by nader-ziada** [comment]
@vincepri @wfernandes any comments of this PR? I'm not of any outstanding changes I should still make. Thanks

**Comment by nader-ziada** [comment]
I don't think the failure in the conformance tests are related to the changes in this PR, will investigate further anyways
